multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCacheManager.java

```diff
@@ -16,21 +32,111 @@
  */
 public class EmbeddedMultimapCacheManager<K, V> implements MultimapCacheManager<K, V> {
 
-   private EmbeddedCacheManager cacheManager;
+   private static final String MULTIMAP_SCOPE = "multimap";
+
+   private final EmbeddedCacheManager cacheManager;
+   private final Map<String, EmbeddedMultimapConfiguration> preConfigured;
+   private final AdvancedCache<ScopedState, EmbeddedMultimapConfiguration> stateCache;
+   private final Map<String, MultimapCache<K, V>> caches = new ConcurrentHashMap<>(16);
 
-   public EmbeddedMultimapCacheManager(EmbeddedCacheManager embeddedMultimapCacheManager){
+   public EmbeddedMultimapCacheManager(EmbeddedCacheManager embeddedMultimapCacheManager) {
       this.cacheManager = embeddedMultimapCacheManager;
+      GlobalConfiguration globalConfiguration = embeddedMultimapCacheManager.getGlobalComponentRegistry().getGlobalConfiguration();
+      MultimapCacheManagerConfiguration configuration = globalConfiguration.module(MultimapCacheManagerConfiguration.class);
+      this.preConfigured = configuration == null
+            ? Collections.emptyMap()
+            : configuration.multimaps();
+      this.stateCache = cacheManager
+            .<ScopedState, EmbeddedMultimapConfiguration>getCache(GlobalConfigurationManager.CONFIG_STATE_CACHE_NAME)
+            .getAdvancedCache();
    }
 
    @Override
    public Configuration defineConfiguration(String name, Configuration configuration) {
-      return cacheManager.defineConfiguration(name, configuration);
+      Configuration registered = cacheManager.defineConfiguration(name, configuration);
+      boolean defined = false;
+      MultimapCacheManagerConfiguration cmc = configuration.module(MultimapCacheManagerConfiguration.class);
+      if (cmc != null) {
+         EmbeddedMultimapConfiguration mc = cmc.multimaps().get(name);
+         if (mc != null) {
+            defineConfiguration(name, mc);
+            defined = true;
+         }
+      }
+
+      if (!defined) {
+         EmbeddedMultimapConfiguration mc = configuration.module(EmbeddedMultimapConfiguration.class);
+         if (mc != null && mc.name().equals(name)) {
+            defineConfiguration(name, mc);
+         }
+      }
+      return registered;
+   }
+
+   @Override
+   public boolean defineConfiguration(String name, Configuration cacheConfiguration, EmbeddedMultimapConfiguration multimapConfiguration) {
+      cacheManager.defineConfiguration(name, cacheConfiguration);
+      return defineConfiguration(name, multimapConfiguration);
+   }
+
+   @Override
+   public boolean defineConfiguration(String name, EmbeddedMultimapConfiguration configuration) {
+      return CompletionStages.join(defineConfigurationAsync(name, configuration));
```

Comment:
This is a possible issue later down the line. We probably want to change the API to return the CompletionStage instead.

Reply:
Yeah, this was something I was struggling with. I only went with the join because the original method here wasn't returning a completion, but I am +1 to return CompletionStage<Boolean>. Going to update.